### DISCUSSION TITLE
投稿したメッセージの送信日時を変更する方法を教えてください

### DISCUSSION THREAD

**Discussion by yuqiao9** [misskey-dev/misskey#15086] [Q&A]
投稿したメッセージの送信日時を変更する方法を教えてください

**Comment by Sayamame-beans** [comment]
意図を詳しく教えていただけますか? (そのような機能に心当たりがないため)

**Comment by yuqiao9** [comment]
> 意図を詳しく教えていただけますか? (そのような機能に心当たりがないため)

2024年12月2日に投稿したメッセージがありますが、2023年1月1日のように日付を変更する方法はありますか。

**Comment by Sayamame-beans** [comment]
ないですね…
(サーバー管理者であれば、ローカルにあるデータをDBから書き換えること自体は可能だとは思いますけれど)

**Comment by yuqiao9** [comment]
> ないですね… (サーバー管理者であれば、ローカルにあるデータをDBから書き換えること自体は可能だとは思いますけれど)

はい、私は管理者ですが、データベースのどのテーブルに対応するデータを変更すべきかはわかりません

**Comment by sakuhanight** [comment]
そのような機能について想定されていないと思います。意図を詳しく教えていただけますか?

**Reply by yuqiao9** [reply]
以前から使っているように見せかけたいので、まず投稿を行い、その後投稿の日時を過去の時間に変更したいと思います

**Comment by AmaseCocoa** [comment]
現在のMisskeyのDBにあるnoteテーブルにはcreatedAtが存在しないので既存のノートのまま日時のみを変更することはできません (IDを変更すればおそらく可能ですがその場合はMisskey側では連合先に配送されない (データベースを直接操作しているため)のでもしも外部にそれを配送するにはrenote等する必要があると思います)

**Reply by yuqiao9** [reply]
私はとても興味があります。毎回ウェブインターフェースを開くたびに、彼はIDに基づいて時間を生成しますか？この時間を保存する場所はないのでしょうか？

**Reply by fruitriin** [reply]
.config/default.conf にidの発番体系が書いてあります
変更していなければ aid が採用されています
aidの生成およびaidから時間の復元アルゴリズムはリポジトリのソースコード読んでください

**Reply by yuqiao9** [reply]
では、具体的にその時間がどのテーブルに保存されているのか教えていただけますか？

**Reply by sakuhanight** [reply]
理論上、aidにより日時を保持していますが、ノートのidを変更することは想定されていません。そのため、あなたの求めている答えを誰も持っていません。
ソースコードを読んでください。

**Reply by AmaseCocoa** [reply]
時間は、noteテーブルのidに保存されたaidやaidxによって保持されています。

また、[fruitriin様の説明](https://github.com/misskey-dev/misskey/discussions/15086#discussioncomment-11455882)に補足すると、[ca00a08](https://github.com/misskey-dev/misskey/commit/ca00a08e6e8ad8636b2002bf8fabcf66d9566015)以降のコミットでは、`.config/example.yml`でaidではなくaidxを利用するように設定されています。aidxやaidは、Misskey Hub上で復元または生成できる[ページ](https://misskey-hub.net/ja/tools/aid-converter/)が用意されています。しかし、[sakuhanight様の説明](https://github.com/misskey-dev/misskey/discussions/15086#discussioncomment-11456613)の通り、Misskeyではノートのidを変更することは想定されていないため、変更を行う際は自己責任で行ってください。

**Reply by eternal-flame-AD** [reply]
正直に言うと、なぜこれが必要なのかまだ理解できませんが、手動で変更するよりも、サーバーのシステム時間を変更してメモを削除し、再作成する方がよいのではないかと思います。

完全には関連していませんが、DBから[スパム分析](https://gist.github.com/eternal-flame-AD/7c9448fcc8627247240aa3e61e61ae30)に使用した作成時間を取得する SQL 関数が書いた。他のことにも役立つかもしれないと思ったので、関連部分を抽出し、参考としてここで共有します。

```sql
BEGIN TRANSACTION READ WRITE;

-- prevent name collision, you can drop the schema cleanly if you no longer want it in the database
CREATE schema IF NOT EXISTS yume;

CREATE OR REPLACE FUNCTION yume.base36_to_int(text) RETURNS bigint AS $$
DECLARE result bigint := 0;
c char;
cd int;
len int;
BEGIN 
    len := length($1);
    FOR i IN 1..len LOOP 
        c := substring($1 FROM i FOR 1);
        cd := ascii(c);
        IF cd >= ASCII('0') AND cd <= ASCII('9') 
            THEN result := result * 36 + cd - ASCII('0');
            ELSE 
                IF cd >= ASCII('a') AND cd <= ASCII('z') 
                    THEN result := result * 36 + cd - ASCII('a') + 10;
                    ELSE RAISE EXCEPTION 'Invalid character in base36 string: %', c;
                END IF;
        END IF;
    END LOOP;
RETURN result;
END;
$$ LANGUAGE plpgsql IMMUTABLE LEAKPROOF;

-- AIDX index to millis from 2000-01-01 (should work on AID too, didn't test)
CREATE OR REPLACE FUNCTION yume.aidx_millis(text) RETURNS bigint AS $$
SELECT yume.base36_to_int(SUBSTRING($1 FROM 1 FOR 8));
$$ LANGUAGE SQL IMMUTABLE LEAKPROOF COST 0.5;

-- AIDX index to timestamp
CREATE OR REPLACE FUNCTION yume.aidx_time(text) RETURNS timestamp AS $$
SELECT make_timestamptz(2000, 1, 1, 0, 0, 0, 'UTC') + 
    yume.aidx_millis($1) / 1000 * interval '1 second' +
    yume.aidx_millis($1) % 1000 * interval '1 millisecond';
$$ LANGUAGE SQL IMMUTABLE LEAKPROOF COST 1;

COMMIT;
```

```sql
-- 最も多く投稿された日付を検索する例
SELECT 
  count(*), 
  date_trunc(
    'day', 
    yume.aidx_time(id)
  ) AS posted_date 
FROM 
  note 
WHERE 
  "userHost" IS NULL 
GROUP BY 
  posted_date 
ORDER BY 
  count DESC 
LIMIT 
  5;

+-------+---------------------+
| count | posted_date         |
|-------+---------------------|
| 49    | 2024-09-29 00:00:00 |
| 43    | 2024-09-30 00:00:00 |
| 40    | 2024-09-18 00:00:00 |
| 37    | 2024-09-25 00:00:00 |
| 36    | 2024-09-17 00:00:00 |
+-------+---------------------+
```
